trainer/trainer.py

```diff
@@ -1,346 +1,605 @@
-        if newseed is None:
-            newseed = self.seed + 1
-        self.seed = newseed
+    def state(self) -> DatasetState:
+        return DatasetState(self.seed, self.line, self.epoch)
+
+    def restore(self, state:DatasetState) -> 'DatasetReader':
+        if self._fh:
+            self._fh.close()
 
-    def _shuffle_(self, inputfile, outputfile):
-        try:
-            #print(self.rng, outputfile, inputfile)
-            check_call([self.rng_filepath, str(self.seed), outputfile, inputfile])
-            #check_call(["/usr/bin/shuf", "-o", outputfile, inputfile])
-        except CalledProcessError as err:
-            print("Error shuffling", inputfile, file=stderr)
-            print(err.cmd, file=stderr)
-            print(err.stderr, file=stderr)
-
-    def _openfile_(self, filepath):
-        self.filehandle = open(filepath, 'rt', encoding="utf-8")
-
-    def get(self) -> Tuple[int, List[str]]:
-        '''Gets the next N lines based on the weight of the dataset. It also reports which
-        epoch it is.
-        When the dataset reaches its end, it automatically takes care of wrapping it'''
-        myepoch = self.epoch
-        retlist: List[str] = []
-        try:
-            for _ in range(int(self.weight*100)):
-                retlist.append(next(self.filehandle))
-                self.linenum = self.linenum + 1
-        except StopIteration:
-            # Update seed and re-shuffle the file UNLESS we have reached the max epoch
-            if self.epoch < self.max_epoch:
-                self.filehandle.close()
-                self._ammend_seed_()
-                self._shuffle_(self.orig, self.shufffile)
-                self._openfile_(self.shufffile)
-                self.epoch = self.epoch + 1
-                self.linenum = 0
-        # Send statistics to the state tracker
-        name = self.filename
-        state = DatasetState(name, self.seed, self.linenum, self.epoch)
-        self.state_tracker.update_dataset(name, state)
-        return (myepoch, retlist)
-
-    def restore_state(self, state: DatasetState) -> None:
-        '''Restores the state from a previous training run'''
-        self.filehandle.close()
         self.seed = state.seed
         self.epoch = state.epoch
-        self._shuffle_(self.orig, self.shufffile)
-        self._openfile_(self.shufffile)
-        i = 0
-        while i < state.line:
-            self.filehandle.readline()
-            i = i + 1
-            self.linenum = self.linenum + 1
+        
+        # Skip forward
+        for _ in range(state.line):
+            next(self)
+
+        return self
+
+    def close(self):
+        if self._fh:
+            self._fh.close()
+
+    def _open(self):
+        # Open temporary file which will contain shuffled version of `cat self.files`
+        fh = TemporaryFile(mode='w+', encoding='utf-8')
+
+        # Shuffle data to the temporary file. Deduplicate on the src side
```

Comment:
Deduplication has been removed. It will have to happen somewhere earlier in the pipeline. Also support for multiple files has been removed from the configuration file to force people to do their deduplication & concatenation beforehand.